modules/staking_pool/src/lib.rs

```diff
@@ -440,8 +487,47 @@ impl<T: Trait> OnNewEra<EraIndex> for Module<T> {
 impl<T: Trait> HomaProtocol<T::AccountId, Balance, EraIndex> for Module<T> {
 	type Balance = Balance;
 
+	/// This function must to be called in `with_transaction_result` scope to
+	/// ensure atomic
 	fn mint(who: &T::AccountId, amount: Self::Balance) -> sp_std::result::Result<Self::Balance, DispatchError> {
-		Self::bond(who, amount)
+		let staking_currency_id = T::StakingCurrencyId::get();
+		T::Currency::ensure_can_withdraw(staking_currency_id, who, amount)
+			.map_err(|_| Error::<T>::StakingCurrencyNotEnough)?;
+
+		let min_bond_ratio = T::MinBondRatio::get();
+		let total_communal_balance = Self::get_total_communal_balance();
+		let communal_bonded = Self::get_communal_bonded();
+
+		// issue liquid currency to who
+		let liquid_amount_to_issue = Self::liquid_exchange_rate()
+			.reciprocal()
+			.unwrap_or_default()
+			.checked_mul_int(amount)
+			.ok_or(Error::<T>::Overflow)?;
+		T::Currency::deposit(T::LiquidCurrencyId::get(), who, liquid_amount_to_issue)?;
+
+		let gap_amount = total_communal_balance
+			.checked_add(amount)
+			.and_then(|n| min_bond_ratio.checked_mul_int(n))
+			.map(|n| n.saturating_sub(communal_bonded))
+			.ok_or(Error::<T>::Overflow)?;
+
+		let amount_to_bond = sp_std::cmp::min(gap_amount, amount);
+		let remainer_to_free = amount.saturating_sub(amount_to_bond);
+
+		// bond
+		Self::bond(who, amount_to_bond)?;
```

Comment:
I think we can just deposit everything into free pool. And let rebalance to do the bonding.